### ISSUE TITLE
Proposal: Extend builtin to include compiler semver

### ISSUE THREAD

**Issue by suirad** [ziglang/zig#1913]
There should be a way to forward and backward prepare the std as well as libraries made with zig for expected changes in the language. The following are the two major examples to explain why:

---
## STD ##
Currently in the std there are parts of the code that are expected to change internally once certain changes happen(i.e copy elision). At the moment, there are points of the std that are annotated with a link to github issue for it to be readdressed. There is the potential for those to be skipped over.
Since most issues are tagged with a version that it is expected to be fixed, an addition to that would be to have those points in code that need to be reworked once the newer version lands announce themselves. 
It could also enable development of workarounds to problems that are expecting larger language changes to work.(i.e version locked networking without coroutines or something)

An illustrative example with psuedo-zig:
```rust
const builtin = @import("builtin");

test "Thing that needs to be reworked after coroutines" {
    // preparing for the coroutine rework to land in 0.4.0
    if (builtin.majVersion != 0 or builtin.minVersion > 3){
        @compileError("Update this test to use coroutines");
    }
    ...
}
```

---

## User libraries ##
Currently libraries built in zig implicitly target whichever version of zig they were built for, but with this change, the code can be more portable and target specific or multiple zig versions in the same codebase.

Psuedo-zig usecase examples:
```rust
const builtin = @import("builtin");

// targeted importing
const mylibfs = switch(builtin.minVersion){
    3 => @import("mylibfs0.3.0.zig"),
    4 => @import("mylibfs0.4.0.zig"),
    else => @compileError("Unsupported Zig version"),
}

// versioned function swapping
const mylibfn = switch(builtin.minVersion){
    3 => mylibfn_without_coroutines,
    4 => mylibfn_with_coroutines,
    else => @compileError("Unsupported Zig version"),
}

fn mylibfn_without_coroutines() !void {
    // implementation without coroutines
}

fn mylibfn_with_coroutines() !void {
    // implementation with coroutines
}

test "use mylibfn" {
    // this fn can be used for both/only zig w/ minver 3 & 4
    mylibfn();
}
```

**Comment by andrewrk** [comment]
Related: #350 and #404

**Comment by andrewrk** [comment]
#6466 is a duplicate of this however there is more activity on the other issue, so closing this one.
